[HLSL] Allow truncation to scalar

llvm-beanz opened the pull request:
HLSL allows implicit conversions to truncate vectors to scalar pr-values. These conversions are scored as vector truncations and should warn appropriately.

This change allows forming a truncation cast to a pr-value, but not an l-value. Truncating a vector to a scalar is performed by loading the first element of the vector and disregarding the remaining elements.

Fixes #102964

llvm-beanz committed:
[HLSL] Allow truncation to scalar

HLSL allows implicit conversions to truncate vectors to scalar
pr-values. These conversions are scored as vector truncations and
should warn appropriately.

This change allows forming a truncation cast to a pr-value, but not an
l-value. Truncating a vector to a scalar is performed by loading the
first element of the vector and disregarding the remaining elements.

Fixes #102964

damyanp commented on clang/lib/Headers/hlsl/hlsl_intrinsics.h:
```diff
@@ -673,9 +673,6 @@ float dot(float3, float3);
 _HLSL_BUILTIN_ALIAS(__builtin_hlsl_dot)
 float dot(float4, float4);
 
-_HLSL_BUILTIN_ALIAS(__builtin_hlsl_dot)
-double dot(double, double);
-
```
Surprised to see this here based on the description. Wonder if we could hand hold readers a bit more to explain why this was removed?

llvm-beanz replied:
Oh, I think this can stay in once this PR lands: https://github.com/llvm/llvm-project/pull/106811

damyanp commented on clang/test/CodeGenHLSL/builtins/dot.hlsl:
```diff
@@ -143,19 +143,3 @@ float test_dot_float3(float3 p0, float3 p1) { return dot(p0, p1); }
 // CHECK: ret float %dx.dot
 float test_dot_float4(float4 p0, float4 p1) { return dot(p0, p1); }
-
-// CHECK:  %dx.dot = call float @llvm.dx.dot2.v2f32(<2 x float> %splat.splat, <2 x float> %1)
-// CHECK: ret float %dx.dot
-float test_dot_float2_splat(float p0, float2 p1) { return dot(p0, p1); }
-
-// CHECK:  %dx.dot = call float @llvm.dx.dot3.v3f32(<3 x float> %splat.splat, <3 x float> %1)
-// CHECK: ret float %dx.dot
-float test_dot_float3_splat(float p0, float3 p1) { return dot(p0, p1); }
-
-// CHECK:  %dx.dot = call float @llvm.dx.dot4.v4f32(<4 x float> %splat.splat, <4 x float> %1)
-// CHECK: ret float %dx.dot
-float test_dot_float4_splat(float p0, float4 p1) { return dot(p0, p1); }
-
-// CHECK: %dx.dot = fmul double %0, %1
-// CHECK: ret double %dx.dot
-double test_dot_double(double p0, double p1) { return dot(p0, p1); }
```
Are these no longer valid cases?  So the `float p0, float2 p1` for example generates an error now, rather than implicitly truncating `p1` to a float?

llvm-beanz replied:
Yes, these are not valid, but also we really shouldn't be testing overload resolution in the intrinsic tests because that just makes the fragile to other changes. The important part of these tests was to verify that the functions codegen to correct code, verifying overload resolution in each of these tests is just excess duplication of test coverage.

bogner commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -2032,26 +2032,42 @@ static bool IsVectorConversion(Sema &S, QualType FromType, QualType ToType,
+        return true;
+      return IsVectorElementConversion(S, FromElTy, ToElTy, ICK, From);
+    }
+    if (FromExtType && nullptr == ToExtType) {
```
Style nit, just using `!` is how this is usually done.
```suggestion
    if (FromExtType && !ToExtType) {
```

bogner reviewed (approved):
Just a few style nitpicks. LG!

llvm-beanz committed:
Merge remote-tracking branch 'origin/main' into cbieneman/scalar-trunc

llvm-beanz committed:
Re-add dot(double,double) overload

llvm-beanz committed:
Update based on PR feedback from @bogner

llvm-beanz commented on clang/lib/Sema/SemaExprCXX.cpp:
```diff
@@ -4828,18 +4830,23 @@ Sema::PerformImplicitConversion(Expr *From, QualType ToType,
+        From = ImpCastExprToType(From, ElType, CK_HLSLVectorTruncation,
+                                 From->getValueKind())
+                   .get();
+      }
```
I changed the naming on this because `ToType` is already used, so this caused shadowing, but otherwise I adopted this structure.

efriedma-quic reviewed (commented):
ExprConstant.cpp currently assumes that an CK_HLSLVectorTruncation can't return a scalar type.

llvm-beanz committed:
Fix constant expression evaluation

This fixes the truncation cast constant evaluation for vectors,
integers and floats.

llvm-beanz commented:
> ExprConstant.cpp currently assumes that an CK_HLSLVectorTruncation can't return a scalar type.

Thank you for catching this! I've updated the PR and included a test that constant evaluates some vector truncations in static asserts.
